src/heightmap.cpp

```diff
@@ -11,6 +12,72 @@
 
 #include <queue>
 
+
+MapLoader::MapLoader()
+{
+    ros::SubscribeOptions map_sub_options = ros::SubscribeOptions::create<nav_msgs::OccupancyGrid>("/map", 1, boost::bind(&MapLoader::mapCallback, this, _1), ros::VoidPtr(), &cb_map);
+    sub_map = nh.subscribe(map_sub_options);
+}
+
+void MapLoader::getMap(nav_msgs::OccupancyGrid& mapRef)
+{
+    if (!initialized)
+    {
+        load();
+    }
+    mapRef = map;
+}
+
+void MapLoader::getMapMetadata(nav_msgs::MapMetaData& metadataRef)
+{
+    if (!initialized)
+    {
+        load();
+    }
+    metadataRef = map.info;
+}
+
+void MapLoader::getMapImage(cv::Mat& imageRef)
+{
+    if (!initialized)
+    {
+        load();
+    }
+    imageRef = mapImage;
+}
+
+void MapLoader::load()
+{
+    ros::Rate r(10);
+    ROS_INFO_STREAM("Waiting for map");
+    while(!initialized && ros::ok())
+    {
+        cb_map.callAvailable();
+        r.sleep();
+    }
+}
+
+void MapLoader::mapCallback(const nav_msgs::OccupancyGrid::ConstPtr& msg)
+{
+    map = *msg;
+    //convert to image
+    mapImage = cv::Mat(map.info.height, map.info.width, CV_8UC1);
+    for (uint row = 0; row < map.info.height; ++row) {
+        for (uint col = 0; col < map.info.width; ++col) {
+            uint index = col + (map.info.height-row) * map.info.width;
+            if (map.data[index] == 0) {
+                mapImage.at<uchar>(row, col) = 255;  // free space
+            } else if (map.data[index] == 100) {
+                mapImage.at<uchar>(row, col) = 0;    // occupied space
+            } else {
+                mapImage.at<uchar>(row, col) = 127;  // door
+            }
```

Comment:
I would suggest to make this a little less strict. Set a threshold where map.data[index] > x results in the pixel becoming a wall and map.data[index] < y results in free space. Now everything not exactly 0 or 100 becomes a door. That will be a lot of doors. 
We may also use the free and occupied thresholds of the map data for this.